### PR TITLE
Implement morm

### PR DESCRIPTION
**Pull request by orkunkl**
This PR implements an improved API for `ModelBucket`
Resolves https://github.com/iov-one/weave/issues/924

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: orm/serial_bucket.go [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
```
Comment (ethanfrey) [inline orm/serial_bucket.go]: Is there a need for an interface here?
Rather we have ONE concrete implementation of this interface.

You can make the interface `serialModuleBucket` and the class `SerialModelBucket`.
This will define the contract and you can enforce that the struct implements the interface with such an compile-time assert. But it is nicer to return a struct and accept an interface, unless you really need to support multiple implementations.

Comment (ethanfrey) [reply orm/serial_bucket.go]: Nevermind, I just realized `migrations.SerialModelBucket`.

But that is a bit confusing with names, one a interface, one a struct

Maybe make `{orm,migrations}.SerialModelBucket` a struct in both cases.
And call the interface: `orm.SerialModelBucketer` (or `ISerialModelBucket` :scream: )

File: orm/serial_bucket.go [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+		}
+		// store the key on the SerialModel
+		dest.Set(reflect.Append(dest, val))
+	}
+	return nil
+
+}
+
+func (smb *serialModelBucket) Create(db weave.KVStore, m SerialModel) error {
+	mTp := reflect.TypeOf(m)
```
Comment (ethanfrey) [inline orm/serial_bucket.go]: I see these reflect checks (7+ lines) in Create and Upsert, and something similar in the load code.

Maybe nicer to pull these out into a named function with all the reflect/type stuff and just use the name here (`enforceType`). It makes the main logic easier to read, as I know many people stop reading when they see `reflect` in go

File: orm/serial_model_bucket_test.go [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+
+func TestSerialModelBucketIndexScanUnique(t *testing.T) {
+	db := store.MemStore()
+
+	b := NewSerialModelBucket("cnts", &CounterWithID{}, WithIndexSerial("counter", lexographicCountIndex, true))
+
+	cnts := []CounterWithID{
+		CounterWithID{Count: 1},
+		CounterWithID{Count: 17},
+		CounterWithID{Count: 93},
```
Comment (ethanfrey) [inline orm/serial_model_bucket_test.go]: can you try this with >255 so the int is now multiple bytes in compressed form
this will show the different between varint encoding and the bigendian (fixed) encoding, as well as littleendian

File: orm/serial_model_bucket_test.go [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+
+func TestSerialModelBucketIndexScanMulti(t *testing.T) {
+	db := store.MemStore()
+
+	b := NewSerialModelBucket("cnts", &CounterWithID{}, WithIndexSerial("counter", lexographicCountIndex, false))
+
+	cnts := []CounterWithID{
+		CounterWithID{Count: 1},
+		CounterWithID{Count: 17},
+		CounterWithID{Count: 3},
```
Comment (ethanfrey) [inline orm/serial_model_bucket_test.go]: same here... try some big - really big - numbers here

File: orm/serial_bucket.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+	// prefix is the kvstore prefix used for all items in the index
+	prefix []byte
+	unique bool
+}
+
+// WithIndexSerial configures the bucket to build an index with given name. All
+// entities stored in the bucket are indexed using value returned by the
+// indexer function. If an index is unique, there can be only one entity
+// referenced per index value.
+func WithIndexSerial(name string, indexer Indexer, unique bool) SerialModelBucketOption {
```
Comment (ethanfrey) [inline orm/serial_bucket.go]: I like these options and that they are only in the Build / New phase. But...

This can actually be called on a bucket after it returns from new.

Maybe add a `fixed bool` flag on `serialModelBucket` that is returned after applying the opts in `NewSerialModelBucket`, and then `WithIndexSerial` panics (or is a no op) if that flag is set.

Addresses the concern of Piotr and Alex of adding Indexes *after* using it.

File: orm/serial_bucket.go [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
```
[the remark this answers is not in the document]
Comment (husio) [reply orm/serial_bucket.go]: I think creating an API of chainable methods (i.e. `bucket.Index(name).PrefixScan(...).One(dest)` is problematic for testing and wrapping. Especially packages like `migration` would have problem with this.

I wonder if all those methods cannot be reduced to this single one though. What if we had only `IndexScan` method? To find entities in the database we would need to write something like this:
```go
err := orm.ByID(store, db, ID, dest) // Calls store.IndexScan underneath.
err := orm.ByIndex(store, db, ID, dest) // Calls store.IndexScan underneath.

iter, err := store.IndexScan(db, indexName, prefix, reverse)
```

It might be slightly less friendly to sometimes provide functionality as a function, other time as a method. The positive side is that the interface is getting smaller and any wrapping is easier. User has to also understand less (look at `ByID` implementation - it is a `IndexScan` call that takes the first result!)

File: orm/serial_bucket.go [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+	// Create saves given SerialModel in the database. Before inserting into
+	// database, SerialModel is validated using its Validate method.
+	// ID field must be unset so auto incremented ID is generated.
+	Create(db weave.KVStore, m SerialModel) error
+
+	// Upsert creates given SerialModel in the database or upserts of given SerialModel
```
[the remark this answers is not in the document]
Comment (husio) [reply orm/serial_bucket.go]: Having those atomic operations would be valid when many clients are accessing the database concurrently. In our case, we run a single threaded program, each event being processed within a transaction, and we use a local database. 
I think in our case, it is fine to call `Has` or `Get` before storing an entity to ensure consistency with put operation. As a consequence we can have a small and simple interface for accessing the data. I believe this is a good thing, because it is easier to remember what function does what.

In my opinion, `Create`, `Upsert` can be reduced to a single method. You can check the intent of the user by investigating ID field.

**Comment by orkunkl** [comment]
During the weekend, I tried my chances with implementing `Index(name)` but as @husio told, It was problematic with `migration` module. As a developer who relatively recently got involved in crypto and blockchain frameworks, I think to have an explicit ID on models makes the API more intuitive, the overhead of getting used to key-value databases is reduced by explicit ID as it feels more `SQLish`. I definitely agree with having multiple bucket implementation makes the API confusing and should be reduced to one.

File: orm/iterator.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	return data.GetRefs(), nil
+}
+
+func (i *indexSerialModelIterator) dbKey(key []byte) []byte {
+	return append(i.bucketPrefix, key...)
+}
+
+func load(key, value, bucketPrefix []byte, dest SerialModel) error {
+	// since we use raw kvstore here, not Bucket, we must remove the bucket prefix manually
```
[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/iterator.go]: Hmm can you explain to me what is the reason of speed up?

File: orm/serial_bucket.go [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+// Objects.
+type SerialModelBucket interface {
+	// GetByID query the database for a single SerialModel instance by ID. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// SerialModel.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given SerialModel type cannot be used to contain stored entity, ErrType
+	// is returned.
+	GetByID(db weave.ReadOnlyKVStore, key []byte, dest SerialModel) error
```
[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/serial_bucket.go]: :+1:

File: orm/serial_bucket.go [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+	// otherwise, it iterates in ascending order.
+	PrefixScan(db weave.ReadOnlyKVStore, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
```
[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/serial_bucket.go]: I like the iterator concept overall. It abstracts this logic into a basic method:

```go
slice := reflect.ValueOf(dest).Elem()
for i := 0; i < slice.Len(); i++ {
	item := slice.Index(i)
}
```

Plus, if you take a look at `ByIndex` methods underlying workings it's obvious that there is a `scan` and iterated job is going on. Logic could be wrapped with an `Iterator` seamlessly

[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/serial_bucket.go]: But also there is the case where users would like the ability to load all objects to the memory with an index so that they can access index + 3rd object? In order to achieve both, we can create a new method with a name like `ByIndexLazy` that returns `SerialModelIterator`. But the necessity of the method should be determined by you as the frequent users and developers of **Weave**. Do you think what do you think?

File: orm/serial_bucket.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/serial_bucket.go]: I like single method approach :)

File: orm/serial_bucket.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (orkunkl) [reply orm/serial_bucket.go]: Hmm... The part I am skeptical about is using `IndexScan` for `ByID` calls result as wrapping of the desired object then unwrapping `SerialModelIterator` to get a single value. Extra steps for every single `ByID` call. This trade-off might reduce performance.

**Comment by orkunkl** [comment]
> @husio  I like the idea of adding an ID to the model. I am not sure of any advantage of this approach over the classic key-value interface. Is there a good example of when this approach wins over the old interface? Right now, this is for me a different way of achieving the same functionality that ModelBucket provides. In my opinion we should not provide more than one way of doing something without a need for it. I believe we should provide beautiful, minimal and useful API that everyone can build upon.

Well, there is not much implementation change internally, you're right on that. But from the perspective of a new learner of Key-Value databases, I find explicit IDs are easier to use as a person who is used SQL. When I started learning using Weave framework, I raised **well I got the models but how to associate these models** question? Providing this functionality will reduce the overhead of learning for the average joe.

File: orm/serial_bucket.go (hunk 5 above)
Comment (orkunkl) [reply orm/serial_bucket.go]: @husio @alpe ?

File: orm/serial_bucket.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply orm/serial_bucket.go]: Fair point with the chainable APIs not being easy to wrap in goland.

Let's leave as it is, just maybe adjust some names.
I don't see that level of `orm.By...` methods being helpful on the API

File: orm/serial_bucket.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply orm/serial_bucket.go]: This is actually one point where I see the use of a global helper function.
I think best the queries that can return multiple values all return **lazy** iterators. This is especially important to allow Limiting the number of responses for public api. (So I can't query `/accounts?prefix` with `""` data and list all accounts, which could be very heavy for the node.).

Some nice helper functions on the orm package: `WithLimit` and `ToSlice`:

```go
// consumes iter until 50 items, then ends early
lim := orm.WithLimit(iter, 50)

// consume all items in the iterator
var data []MyModel
err := orm.ToSlice(lim, &data)
```

In this model, `orm.ToSlice(bucket.IndexScan(db, index, prefix, reverse), &data)` would do what `bucket.ByIndex(db, index, prefix, reverse, &data)` does now, but in a more composible way (eg. easy to add the limit filter there without overwhelming the argument list)

**Comment by ethanfrey** [comment]
@husio you will not like this, but I think your API small interface purity is not good in this cases.

We are still quite minimal (serializable, validatable, and with id) and good to keep an eye on this, but I truly feel this makes the usage much nicer and is the continuation of where were going with MdoelBucket and IDGenBucket and all that. Keeping those interfaces and wrapping and composability is a nice aim, but also hit it's limit IMO. See discussion on the VersioningBucket.

Those other buckets can co-exist. I would like this interface to be easy to use for novices first, and perfect idiomatic go second. Many of the comments (like avoiding chains, and only exposing methods that are truly necessary) are good points. 

eg. `Get` and `Iterator` are different use cases with much different overhead and should not be reduced to one method for some purity sake. If we can treat the result of `PrefixScan` and `IndexScan` the same way... yes, I agree that does simplify the API.

In short, good to push for smaller, cleaner interfaces, but I'm going to take a strong stand with Orkun here, aim is easy to use for novice devs, without an over bloated API (and this includes nice go docs and example queries).

File: orm/serial_bucket.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (husio) [reply orm/serial_bucket.go]: I think @ethanfrey understands my suggestion well.

Having all functions that can return more than one model result return an iterator is in my opinion a nice unification. With proposed by Ethan helpers (`orm.WithLimit` and `orm.ToSlice`), we provide the necessary tooling to work with the iterators. 

This approach reminds me of `io.Reader` interface and [`LimitReader`](https://golang.org/pkg/io/#LimitedReader) and [`ReadAll`](https://golang.org/pkg/io/ioutil/#ReadAll) functions.

File: orm/serial_bucket.go (hunk 5 above)
Comment (husio) [reply orm/serial_bucket.go]: This is possible, because `SerialModelBucketOption` takes a shortcut on what it accepts. "Usually" the argument would be a function that accepts a private configuration object, not the retruned by `NewXxx` function. For example:
```go
type options struct {                                                          
        index string                                                           
}                                                                              
                                                                               
type Option func(*options)                                                     
                                                                               
func WithIndexSerial(name string) Option {                                     
        return func(o *options) {          
                // No direct access to the bucket instance here. 
                // Only the options structure can be modified.                                    
                o.index = name                                                 
        }                                                                      
}                                                                              
                                                                               
func NewModel(args ...Option) *serialModelBucket {                             
        var o options                                                          
        for _, fn := range args {                                              
                fn(&o)                                                         
        }                                                                      
                                                                               
        var bucket serialModelBucket                                           
               
        // Inspect the options structure to decide which functionality was configured.                                                                
        if o.index {                                                           
                // stuff                                                       
                // bucket.index.xxx                                            
        }                                                                      
                                                                               
        return &bucket                                                         
}   
```

There is no `options` entity in this implementation. Instead, the bucket is modified directly. 

I am fine with both implementations - direct access and through options structure proxy. Implementation with the proxy is safer as it cannot change the model instance directly.

File: orm/serial_bucket.go [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetID/SetID are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetID() []byte
+	SetID([]byte) error
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply orm/serial_bucket.go]: Interesting approach.

This does assume that the primary key is always prepared and I think it may push some more work to the users of the API.

Note that `GetID` is auto-generated from protobuf, so this was only one method in either case.

I do like the rename to `PrimaryKey` over `ID` as well.

**Comment by orkunkl** [comment]
It's been a long time since this PR is WIP. Let's finish it.
To summarize all this discussion:
1 - [] add a fixed bool flag on serialModelBucket that is returned after applying the opts in NewSerialModelBucket, and then WithIndexSerial panics (or is a no op) if that flag is set.
2 - [] Reduce bucket implementations in future
3 - [] Queries should return multiple values all. return lazy iterators.
4 - [] Add helper functions on the orm package: `WithLimit` and `ToSlice`
5 - [] Reduce `Create` and `Upsert` to a single method
6 - [] Rename `ID` to `PrimaryKey`
7 - [] Add a `fixed bool` flag on `serialModelBucket` that is returned after applying the opts in `NewSerialModelBucket`, and then `WithIndexSerial` panics (or is a no op) if that flag is set.
@husio anything missing?

**Comment by husio** [comment]
> -  Reduce bucket implementations in future

In the future doesn't count :stuck_out_tongue:
If you want to do it, do it now. Otherwise, no need to mention it.

I don't quite understand the first and the last point.

Everything else is :ok_hand:

**Comment by husio** [comment]
All I am saying is that `in the future` is almost always just another way of saying `never`.

**Comment by orkunkl** [comment]
@husio Then I took back my word 🙊

File: orm/serial_bucket.go [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+			return errors.Wrap(err, "ID sequence")
+		}
+	}
+
+	obj := NewSimpleObj(key, m)
+	if err := smb.b.Save(db, obj); err != nil {
+		return errors.Wrap(err, "cannot create in the database")
+	}
+	// after serialization, return original/generated key on SerialModel
+	if err := m.SetPrimaryKey(key); err != nil {
```
Comment (alpe) [inline orm/serial_bucket.go]: req: you should store the key with the model before persisting the object. It will save you all the work in setting it later after load. Any modification on objects after load may get you into trouble with Proofs.

File: orm/serial_bucket.go [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// Create saves given SerialModel in the database.
+	//If ID field is provided uses it otherwise generates an auto-incremented key.
+	// Before inserting into database, SerialModel is validated using Validate method.
+	Create(db weave.KVStore, m SerialModel) error
```
Comment (alpe) [inline orm/serial_bucket.go]: req: There is no update function defined. With `Create` I have the assumption that it would fail when an id exists already but this is not the case. Therefore rename it to `Save`, `Store`,  `Put`, ...

File: orm/serial_bucket.go (hunk 12 above)
Comment (ethanfrey) [reply orm/serial_bucket.go]: Hmm... interesting. I guess it does waste a bit of space, but then makes the whole module much clearer.

The object is just like a model from sql / orm land - all fields.
Then we calculate the PrimaryKey as well as each index from it.

I think this adds a bit of space overhead, but for items with auto-increment primary key, not too bad. I wonder about cases when the key is already 20/32 bytes (eg. addresses). But in general, a very nice simplification.

File: orm/serial_bucket.go (hunk 13 above)
Comment (ethanfrey) [reply orm/serial_bucket.go]: I believe there was once separate cases for Create and Update - one required no key, the other with a key. But it was requested to simplify. I do kind of like that separation to make the code clearer when using it, but it does add some overhead.

File: orm/serial_bucket.go [hunk 14]
```diff
@@ -258,6 +258,7 @@ func (smb *serialModelBucket) IndexScan(db weave.ReadOnlyKVStore, indexName stri
 }
 
 func (smb *serialModelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination SerialModelSlicePtr) error {
+	// TODO refactor this function to return an iterator
```
Comment (husio) [inline orm/serial_bucket.go]: Please don't leave _TODO_ in code. Instead, explain **why** it is implemented like it is and how could it be done better.

Leaving `TODO do X` never motivates to address this comment. As a result, this comment will stay here forever providing no information beyond an impression that this code is half done.

File: orm/serial_bucket.go [hunk 15]
```diff
@@ -258,7 +258,10 @@ func (smb *serialModelBucket) IndexScan(db weave.ReadOnlyKVStore, indexName stri
 }
 
 func (smb *serialModelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination SerialModelSlicePtr) error {
-	// TODO refactor this function to return an iterator
+	// TODO Refactor this function to return lazy iterator
+	// get indexed loads all the objects to objs where
+	// the iterator wrapping must happen. Should be taken care
+	// serial model bucket is optimized by moving bucket logic here
```
Comment (husio) [inline orm/serial_bucket.go]: I change my mind, sorry for that.

Right now, the function loads all values and returns all of them at once. It returns only an error. I think this comment should be removed completely, because described change will not happen. If this bucket implementation will be used, it is easier to introduce another method instead and maybe remove this one.

Changing notation of a method is hard if you use composition. You cannot use whatever notation you like, you must use what is expected.

File: orm/counter.go [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package orm
+
+var _ Model = (*CounterWithID)(nil)
+
+// SetPrimaryKey is a minimal implementation, useful when the ID is a separate protobuf field
+func (c *CounterWithID) SetPrimaryKey(pk []byte) error {
+	c.PrimaryKey = pk
+	// TODO possible errors
```
Comment (husio) [inline orm/counter.go]: Please address or remove TODO.

File: orm/counter.go [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+	return nil
+}
+
+// Validate is always succesful
+func (c *CounterWithID) Validate() error {
+	// TODO possible errors
```
Comment (husio) [inline orm/counter.go]: Please address or remove TODO.

File: orm/errors.go [hunk 18]
```diff
@@ -8,3 +8,7 @@ import (
 
 // ErrInvalidIndex is returned when an index specified is invalid
 var ErrInvalidIndex = errors.Register(100, "invalid index")
+
+// ErrBucketFixed is returned when already initialized bucket is tried
+// to be indexed again
+var ErrBucketFixed = errors.Register(101, "bucket fixed")
```
Comment (husio) [inline orm/errors.go]: Would it be better to call it `ErrBucket`, `ErrConflict` or `ErrIndex`? I think the _Fixed_ part does not belong here.

File: orm/helpers.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave/errors"
+)
+
+type limitedIterator struct {
+	// n is count of remaining elements
+	n int
```
Comment (husio) [inline orm/helpers.go]: I think it would be better to call this variable `remaining` and remove the comment.

File: orm/helpers.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+	// iterator is the underlying iterator
+	iter SerialModelIterator
+}
+
+var _ SerialModelIterator = (*limitedIterator)(nil)
+
+func (l *limitedIterator) LoadNext(dest SerialModel) error {
+	if l.n > 0 {
+		l.n--
+		return l.iter.LoadNext(dest)
```
Comment (husio) [inline orm/helpers.go]: I think you should not decrement the counter if the iterator returned an error. The current approach can produce an unexpected result:
```
it := Limit(bucket, 2)
val1, err := next(it) // ok
_, err := next(it) // temporary error
val2, err := next(it) // ErrIteratorDone, although I expect a second value

File: orm/helpers.go [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	return errors.Wrap(errors.ErrIteratorDone, "iterator limit reached")
+}
+
+func (l *limitedIterator) Release() {
+	l.iter.Release()
+}
+
+func WithLimit(iter SerialModelIterator, limit int) (SerialModelIterator, error) {
```
Comment (husio) [inline orm/helpers.go]: I think this function could be improved.

Rename `WithLimit` to `LimitIterator`, to match [`io.LimitReader`](https://golang.org/pkg/io/#LimitReader). I think `Limit` would also be a good name.

Removing the error would make much nicer API. It would be ```Limit(SerialModelIterator, int) SerialModelIterator``` and the caller does not have to test error. Giving a limit value less than 1 makes little sense, but it is not invalid. Creating a limit of 0 or less will produce a valid iterator that always returns `ErrIteratorDone`.

File: orm/helpers.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	if dest.IsNil() {
+		return errors.Wrap(errors.ErrImmutable, "got nil pointer")
+	}
+	dest = dest.Elem()
+	if dest.Kind() != reflect.Slice {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of SerialModels")
+	}
+
+	// Allocate memory depending on the type of the destination
```
Comment (husio) [inline orm/helpers.go]: This comment might be misleading, because below code does not allocate memory. It only determines the type of the model.

File: orm/serial_bucket.go [hunk 23]
```diff
@@ -0,0 +1,373 @@
...
+		b:          b,
+		idSeq:      b.Sequence("id"),
+		model:      tp,
+		bucketName: name,
+	}
+	for _, fn := range opts {
+		fn(smb)
+	}
+
+	// set fixed
```
Comment (husio) [inline orm/serial_bucket.go]: This comment has no value.

File: orm/serial_bucket.go (hunk 15 above)
Comment (orkunkl) [reply orm/serial_bucket.go]: > Changing notation of a method is hard if you use composition. You cannot use whatever notation you like, you must use what is expected.

Can you explain the `composition here`

Comment (husio) [reply orm/serial_bucket.go]: In Go, preffered is [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).

As an example, `migration` package does not inherit from `ModelBucket`. Instead, it enrich `ModelBucket` functionality by providing the very same interface with a custom implementation.
The above would not be possible if the notation of the functions would change. You would have to update all places where it is used.
